### PR TITLE
Added the option to write the measurement errors to the secondary output file

### PR DESCRIPTION
**Pull request by garethethomas**
It can be quite useful to see what values for measurement error are actually being used in the retrieval, so this adds the option to request the diagonals of the Sy covariance matrix be output to the seondary.nc file.

### CODE REVIEW

File: src/get_spixel.F90 [hunk 1]
```diff
@@ -342,6 +354,16 @@ subroutine Get_SPixel(Ctrl, SAD_Chan, SAD_LUT, MSI_Data, RTM, SPixel, status)
    call Get_X(Ctrl, SPixel, MSI_Data, status)
 !  if (status /= 0) go to 99 ! Skip further data reading
 
+   ! It seems that it is possible that an invalid pixel, with no
+   ! defined measurement, can reach this point without causing a
+   ! SPixel status. Check for this condition here
+   if (SPixel%Ind%Ny == 0) then
```
Comment (adamcpovey) [inline src/get_spixel.F90:360]: While I'm not surprised this is possible, someone should track down what the failure mode is and fix that too.

File: src/prepare_output_secondary.F90 [hunk 2]
```diff
@@ -330,6 +331,44 @@ subroutine prepare_output_secondary(Ctrl, i, j, MSI_Data, SPixel, Diag, &
...
+      !   allocated.
+      ! * If the Sy array has the correct dimension. If we have missing
+      !   channels and skipped pixels, it is also possible at Ny ends up
+      !   being larger than Sy, at this point in the code.
+      if (SPixel%Ind%Ny .gt. 1 .and. size(SPixel%Sy) .gt. 1 .and. &
+           SPixel%Ind%Ny*SPixel%Ind%Ny .le. size(SPixel%Sy)) then
+         do k=1,SPixel%Ind%Ny
+            !write(*,"(e9.2)", advance="no") sqrt(SPixel%Sy(k,k))
+            call prepare_short_packed_float( &
+                 sqrt(SPixel%Sy(k,k)), output_data%Sy(i,j,k), &
```
Comment (adamcpovey) [inline src/prepare_output_secondary.F90:355]: This doesn't map channels between SPixel and Ctrl and so is outputting in the wrong place. I'll work out a fix momentarily.

**Review by adamcpovey — CHANGES_REQUESTED** [review]
I'll make some whitespace tidying and work out the channel indexing

**Comment by adamcpovey** [comment]
After 20 minutes I worked out how to push changes to your branch and update this thing! @garethethomas please check that those still address the problems you were having (I went looking for the causes of the missed status flag rather than circumvent the whole system).

Further, I am beginning to question a bunch of stuff in here:

- Why are we allocating SPixel arrays each time when they just get copied into local arrays in `invert_marquadt()`? We probably save a bunch of processing time by not bothering to reallocate.
- Given the output arrays are initialised to fill values, why do we call `prepare_output_XXX()` for skipped retrievals? Surely doing nothing outputs what we want (namely nothing)?

**Comment by garethethomas** [comment]
I agree with your inclusion of `SPixel%spixel_y_to_ctrl_y_index` - from what I've seen it isn't generally necessary (it's a one-to-one match), but I guess this might not _always_ be the case!
Regarding your other change - it _should_ be ok, so long as `SPixel%Ny = 0` means that the do loop at line 338 of prepare_output_secondary.F90  doesn't get executed. Is that how Fortran works?

I was also surprised at the fact that all of the SPixel arrays are reallocated for every pixel - like you say, it seems rather inefficient. "Fixing" this would be a big job though!

**Comment by adamcpovey** [comment]
`SPixel%spixel_y_to_ctrl_y_index` is mostly important for the cloud retrieval, as one of the effective radius channels is _always_ missing when multiple are present.

To my knowledge, Fortran checks the validity of the for condition at the start of the loop so `do i=1,0` should immediately terminate.

**Comment by adamcpovey** [comment]
I never managed to work out how to get sensible errors out of flex. You can get it to at least throw a line number and slightly more information (almost always "syntax", but others can happen) by uncommenting the last two lines of `config/arch.inc`

**Pull request merged**